### PR TITLE
Webpack 4

### LINKED ISSUE #241: [Question] Any interest in using Webpack v4 ?
### Is this a bug report?

No

Is there any interest in upgrading to Webpack v4? Are there any known issues with upgrading to v4? If it is of interest, I or someone else can likely provide a hand in upgrading.

### PR DESCRIPTION
**Pull request by halfzebra**
This PR enables Webpack 4 support.

### CLI
- [case-sensitive-paths-webpack-plugin](https://github.com/Urthen/case-sensitive-paths-webpack-plugin) added for development mode.
- Removes `yarn.lock`
- Replaces [assets-webpack-plugin](https://github.com/kossnocorp/assets-webpack-plugin) witn [webpack-manifest-plugin](https://github.com/danethurber/webpack-manifest-plugin)
- Replaces [extract-text-webpack-plugin](https://github.com/webpack-contrib/extract-text-webpack-plugin) with [mini-css-extract-plugin](https://github.com/webpack-contrib/mini-css-extract-plugin)

### Dev
- Enables [lint-staged](https://github.com/okonet/lint-staged) with Prettier

Dependencies: https://github.com/facebook/create-react-app/pull/4077

While it is possible to update the [InterpolateHtmlPlugin](https://github.com/facebook/create-react-app/blob/next/packages/react-dev-utils/InterpolateHtmlPlugin.js) and use a temporary replacement, I think it makes sense to wait for CRA release and release CEA with all the updates.

Closes #241

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -30,29 +29,32 @@
...
     "elm-hot-loader": "0.5.4",
     "elm-test": "^0.18.12",
-    "elm-webpack-loader": "^4.4.0",
-    "extract-text-webpack-plugin": "^3.0.2",
-    "file-loader": "^1.1.6",
+    "elm-webpack-loader": "^4.5.0",
+    "file-loader": "^1.1.11",
     "fs-extra": "^5.0.0",
-    "html-webpack-plugin": "^2.30.1",
+    "html-webpack-plugin": "^3.0.7cd tem  ",
```
Comment (pruett) [inline package.json]: ⚠️

**Comment by halfzebra** [comment]
@pruett thanks for pointing out! Fixed 👍

File: config/webpack.config.prod.js [hunk 2]
```diff
@@ -249,10 +275,30 @@ module.exports = {
...
+    // The public URL is available as %PUBLIC_URL% in index.html, e.g.:
+    // <link rel="shortcut icon" href="%PUBLIC_URL%/favicon.ico">
+    // In production, it will be an empty string unless you specify "homepage"
+    // in `package.json`, in which case it will be the pathname of that URL.
+    new InterpolateHtmlPlugin(env.raw),
+    // Makes some environment variables available to the JS code, for example:
+    // if (process.env.NODE_ENV === 'production') { ... }. See `./env.js`.
+    // It is absolutely essential that NODE_ENV was set to production here.
+    // Otherwise React will be compiled in the very slow development mode.
+    new webpack.DefinePlugin(env.stringified),
```
Comment (littleStudent) [inline config/webpack.config.prod.js]: `new webpack.DefinePlugin(env.stringified)`, is added twice

Comment (halfzebra) [reply config/webpack.config.prod.js]: Good catch, thanks!

**Comment by littleStudent** [comment]
since https://github.com/facebook/create-react-app/pull/4077 is merged now.
Can this PR move forward? Or are there any other dependencies?

**Comment by halfzebra** [comment]
None of the dependencies were released to NPM, the progress can be tracked here https://github.com/facebook/create-react-app/issues/3815

**Comment by halfzebra** [comment]
Create Elm App with Webpack 4 is available at `npm i create-elm-app@next`

Please let me know if you are experiencing any issues!

**Pull request merged**